kasa/aestransport.py

```diff
@@ -185,34 +187,41 @@ async def perform_login(self):
         self._handle_response_error_code(resp_dict, "Error logging in")
         self._login_token = resp_dict["result"]["token"]
 
-    async def perform_handshake(self):
-        """Perform the handshake."""
-        _LOGGER.debug("Will perform handshaking...")
-        _LOGGER.debug("Generating keypair")
-
-        self._handshake_done = False
-        self._session_expire_at = None
-        self._session_cookie = None
-
-        url = f"http://{self._host}/app"
-        key_pair = KeyPair.create_key_pair()
+    async def _generate_request_body(self) -> AsyncGenerator:
+        """Generate the request body and return an ascyn_generator.
 
+        This prevents the key pair being generated unless a connection
+        can be made to the device.
+        """
+        self._key_pair = KeyPair.create_key_pair()
         pub_key = (
             "-----BEGIN PUBLIC KEY-----\n"
-            + key_pair.get_public_key()
+            + self._key_pair.get_public_key()  # type: ignore[union-attr]
             + "\n-----END PUBLIC KEY-----\n"
         )
         handshake_params = {"key": pub_key}
         _LOGGER.debug(f"Handshake params: {handshake_params}")
-
         request_body = {"method": "handshake", "params": handshake_params}
-
         _LOGGER.debug(f"Request {request_body}")
+        yield json_dumps(request_body).encode()
+
+    async def perform_handshake(self):
+        """Perform the handshake."""
+        _LOGGER.debug("Will perform handshaking...")
+        _LOGGER.debug("Generating keypair")
 
+        self._key_pair = None
+        self._handshake_done = False
+        self._session_expire_at = None
+        self._session_cookie = None
+
+        url = f"http://{self._host}/app"
+        # Device needs the content length or it will response with 500
+        headers = {**self.COMMON_HEADERS, "Content-Length": "318"}
         status_code, resp_dict = await self._http_client.post(
             url,
```

Comment:
Hardcoding a const here feels error prone, but the pub key should always be of the same length so I suppose it's fine. I'd still create a const variable on top of the file for this to avoid bare magic numbers.

Reply:
Done